csrc/multidevice/pipeline.cpp

```diff
@@ -0,0 +1,307 @@
+  void buildPipelineVals() {
+    for (auto stage_desc : pipeline_->descriptor().stageDescriptors) {
+      // Create a PipelineVal for each stage's input
+      for (auto val : StageInputsDesc[stage_desc].vector()) {
+        auto pVal =
+            IrBuilder::create<PipelineVal>(pipeline_->as<IrContainer>(), val);
+        valToPipelineValInputOfStage[val] = pVal;
+        // If the Val is an Input of the original Fusion,
+        // then add the newly created PipelineVal as an input of the pipeline
+        if (isGlobalInput(val)) {
+          pipeline_->addInput(pVal);
+          TORCH_INTERNAL_ASSERT(
+              stage_desc->mesh.size() == 1,
+              "A global input must belong to a stage which mesh is of size 1");
+        } else {
+          // if the Val is a stage input but not a global input, it must be
+          // defined by a "Set" operation
+          TORCH_INTERNAL_ASSERT(
+              (val->definition()->isA<LoadStoreOp>()) &&
+                  (val->definition()->as<LoadStoreOp>()->opType() ==
+                   LoadStoreOpType::Set),
+              "A Val that is the input of a stage must be defined by a LoadStoreOp expression of type Set"
+              "but here the definition is " +
+                  val->definition()->toString());
+        }
+      }
+      // Create a PipelineVal for each stage's output
+      for (auto val : StageOutputsDesc[stage_desc].vector()) {
+        auto pVal =
+            IrBuilder::create<PipelineVal>(pipeline_->as<IrContainer>(), val);
+        valToPipelineValOutputOfStage[val] = pVal;
+        if (isGlobalOutput(val)) {
+          pipeline_->addOutput(pVal);
+        }
+      }
+    }
+  }
+
+  // Build the PipelineStage IR of the Pipeline
+  // An PipelineStage is created for each Stage of the Fusion
+  void buildPipelineStages() {
+    for (auto& stage_desc : pipeline_->descriptor().stageDescriptors) {
+      // containers for storing the I/O of the PipelineStage
+      ValSet ins, outs;
+      for (auto& val : StageInputsDesc[stage_desc]) {
+        ins.pushBack(valToPipelineValInputOfStage[val]);
+      }
+      for (auto& val : StageOutputsDesc[stage_desc]) {
+        outs.pushBack(valToPipelineValOutputOfStage[val]);
+      }
+      auto stage = IrBuilder::create<PipelineStage>(
+          pipeline_->as<IrContainer>(), stage_desc, ins, outs);
+      stages.push_back(stage);
+    }
+  }
+
+  // Build the PipelineCommunication IR of the Pipeline
+  // A PipelineCommunication is created for each stage's input
+  void buildPipelineCommunications() {
+    for (auto& stage : stages) {
+      for (auto& pVal : stage->inputs()) {
+        std::vector<Val*> ins, outs;
+        auto val = pVal->as<PipelineVal>()->getOriginalVal();
+        if (isGlobalInput(val)) {
+          continue;
+        }
+        outs.push_back(pVal);
+        for (auto& producer : val->definition()->inputs()) {
+          ins.push_back(valToPipelineValOutputOfStage[producer]);
```

Comment:
I'd suggest to use `at` instead of `[]` just in case.

Reply:
done. Same thing for all the occurrences of `[]` or just this one ?